docs/architecture/adr-030-msg-auth-module.md

```diff
@@ -0,0 +1,222 @@
+  // authorizations granted to the grantee. Each message should have only
+  // one signer corresponding to the granter of the authorization.
+  rpc ExecAuthorized(MsgExecAuthorized) returns (MsgExecAuthorizedResponse)
+
+
+  // RevokeAuthorization revokes any authorization corresponding to the provided method name on the
+  // granter's account that has been granted to the grantee.
+  rpc RevokeAuthorization(MsgRevokeAuthorization) returns (MsgRevokeAuthorizationResponse);
+}
+
+message MsgGrantAuthorization{
+  string granter = 1;
+  string grantee = 2;
+  google.protobuf.Any authorization = 3 [(cosmos_proto.accepts_interface) = "Authorization"];
+  google.protobuf.Timestamp expiration = 4;
+}
+
+message MsgExecAuthorized {
+    string grantee = 1;
+    repeated google.protobuf.Any msgs = 2;
+}
+
+message MsgRevokeAuthorization{
+  string granter = 1;
+  string grantee = 2;
+  string method_name = 3;
+}
+```
+
+### Router Middleware
+
+The `msg_authorization` `Keeper` will expose a `DispatchActions` method which allows other modules to send `ServiceMsg`s
+to the router based on `Authorization` grants:
+
+```go
+type Keeper interface {
+  DispatchActions(ctx sdk.Context, grantee sdk.AccAddress, msgs []sdk.ServiceMsg) sdk.Result`
+}
+```
+
+This allows the functionality provided by `msg_authorization` to be used for future inter-module object capabilities
+permissions as described in https://github.com/cosmos/cosmos-sdk.
+
+### CLI
+
+#### `--send-as` Flag
+
+When a CLI user wants to run a transaction on behalf of another account using `MsgExecAuthorized`, they
+can use the `--send-as` flag to specify the granter's account but signed by the grantee's account as
+specified by the `--from` flag. For instance `gaiacli tx gov vote 1 yes --from mykey --send-as cosmos3thsdgh983egh823`
+would send a transaction like this:
+
+```go
+MsgExecAuthorized {
+  Grantee: mykey,
+  Msgs: []sdk.SericeMsg{
+    ServiceMsg {
+      MethodName:"/cosmos.gov.v1beta1.Msg/Vote"
+      Request: MsgVote {
+	    ProposalID: 1,
+	    Voter: cosmos3thsdgh983egh823
+	    Option: Yes
+      }
+    }
+  }
+}
+```
+
+#### `tx grant <grantee> <authorization> --from <granter>`
+
+This CLI command will send a `MsgGrantAuthorization` transaction. `authorization` should be encoded as
+JSON on the CLI.
+
+#### `tx revoke <grantee> <method-name> --from <granter>`
+
+This CLI command will send a `MsgRevokeAuthorization` transaction.
+
+### Built-in Authorizations
+
+#### `SendAuthorization`
+
+```proto
+// SendAuthorization allows the grantee to spend up to spend_limit coins from
+// the granter's account.
+message SendAuthorization {
+  repeated cosmos.base.v1beta1.Coin spend_limit = 1;
+}
+```
+
+#### `GenericAuthorization`
+
+```proto
+// GenericAuthorization gives the grantee unrestricted permissions to execute
+// the provide method on behalf of the granter's account.
+message GenericAuthorization {
+  string method_name = 1;
```

Comment:
We need to use a different name for`method_name`. There is already one method is present with the same name in the `Authorization` interface.

Reply:
You can just disable generating getters with a `gogoproto` extension if that's the issue.